### PR TITLE
[SPARK-50077][SQL] Introduce a new pattern object for LogicalRelation to help avoiding default full params pattern

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to introduce a new pattern object to extract BaseRelation and CatalogTable from LogicalRelation, and update existing pattern matches with LogicalRelation with the new pattern.

This PR also adds the information of pattern object and encourage devs to use it.

### Why are the changes needed?

I found it very difficult to add a param into LogicalRelation, because a bunch of places use pattern matching with LogicalRelation, and all of them are using default pattern with specifying all params (despite wildcards).

I looked into the patterns and realized most of the patterns just need relation and catalog table inside the LogicalRelation. output and isStreaming are used infrequently. That said, introducing a new pattern object extracting BaseRelation and CatalogTable from LogicalRelation would cover most of the patterns we have for LogicalRelation, and this would help a lot to introduce a new param into LogicalRelation in future.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

N/A (Existing tests would cover it)

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
cc. @cloud-fan Would you mind taking a look? Please cc. to others if you are busy with other stuff. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 1]
```diff
@@ -102,3 +109,18 @@ object LogicalRelation {
...
   }
 }
+
+/**
+ * Extract the [[BaseRelation]] and [[CatalogTable]] from [[LogicalRelation]]. This also provides
+ * [[LogicalRelation]] itself, to cover the case where other params are needed (but not required
+ * to have specific value to match).
+ */
+object RelationAndCatalogTable {
+  def unapply(plan: LogicalPlan): Option[(LogicalRelation, BaseRelation, Option[CatalogTable])] = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala:119]:
```suggestion
  def unapply(plan: LogicalRelation): Option[(BaseRelation, Option[CatalogTable])] = {
```

How about this? Then the caller can be simplified `l @ RelationAndCatalogTable(fsRelation: HadoopFsRelation, _)`

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 2]
```diff
@@ -102,3 +109,18 @@ object LogicalRelation {
     LogicalRelation(relation, toAttributes(schema), Some(table), false)
   }
 }
+
+/**
+ * Extract the [[BaseRelation]] and [[CatalogTable]] from [[LogicalRelation]]. This also provides
+ * [[LogicalRelation]] itself, to cover the case where other params are needed (but not required
+ * to have specific value to match).
+ */
+object RelationAndCatalogTable {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala:118]: `LogicalRelationWithTable`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala (hunk 1 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala:119]: Yeah if this work, then looks good, especially with your suggested name. The new name implies they could get the LogicalRelation via `l @ LogicalRelationWithTable(...)`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala (hunk 2 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala:118]: Don't have a strong voice for the name; it looks better coupled with your above comment. Will rename.

**Comment by HeartSaVioR** [comment]
Thanks @cloud-fan for your review! I've addressed your comments. PTAL.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 3]
```diff
@@ -102,3 +109,18 @@ object LogicalRelation {
     LogicalRelation(relation, toAttributes(schema), Some(table), false)
   }
 }
+
+/**
+ * Extract the [[BaseRelation]] and [[CatalogTable]] from [[LogicalRelation]]. This also provides
+ * [[LogicalRelation]] itself, to cover the case where other params are needed (but not required
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala:115]: shall we update this comment?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala [hunk 4]
```diff
@@ -102,3 +109,18 @@ object LogicalRelation {
...
+ * to have specific value to match).
+ */
+object LogicalRelationWithTable {
+  def unapply(plan: LogicalRelation): Option[(BaseRelation, Option[CatalogTable])] = {
+    plan match {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala:120]: no need to do pattern match now

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala (hunk 3 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/LogicalRelation.scala:115]: Thanks for catching this, forgot to change.

**Comment by HeartSaVioR** [comment]
I've addressed the latest comments. Thanks! Merging to master.

**Pull request closed without merging**